Land Lockdown Mode features in Open Source

brentfulgham opened the pull request:
#### 73157b607c911ec3c5e16920f1dbbf44a3ac55ae
<pre>
Land Lockdown Mode features in Open Source
<a href="https://bugs.webkit.org/show_bug.cgi?id=242767">https://bugs.webkit.org/show_bug.cgi?id=242767</a>

Reviewed by Chris Dumez.

* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
(-[WKWebpagePreferences lockdownModeEnabled]):
(-[WKWebpagePreferences _setLockdownModeEnabled:]):
(-[WKWebpagePreferences _willChangeCaptivePortalMode]):
(-[WKWebpagePreferences _didChangeCaptivePortalMode]):
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _presentCaptivePortalModeAlertOnce]):
(-[WKWebView _presentCaptivePortalModeAlertIfNeeded]):

Canonical link: <a href="https://commits.webkit.org/252540@main">https://commits.webkit.org/252540@main</a>
</pre>

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
```diff
@@ -554,15 +554,34 @@ - (_WKWebsiteModalContainerObservationPolicy)_modalContainerObservationPolicy
+    return _websitePolicies->captivePortalModeEnabled();
+}
+
+- (void)_setLockdownModeEnabled:(BOOL)lockdownModeEnabled
```
I know you merely moved the code but I don't get why we have a leading underscore here.

brentfulgham replied:
Fixed! I'm guessing it wouldn't have worked with the underline (or at least not done what we expected).

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
```diff
@@ -554,15 +554,34 @@ - (_WKWebsiteModalContainerObservationPolicy)_modalContainerObservationPolicy
-#else
+- (BOOL)lockdownModeEnabled
+{
+    return _websitePolicies->captivePortalModeEnabled();
```
Shouldn't this be protected with `#if ENABLE(LOCKDOWN_MODE_API)`?

brentfulgham replied:
Hmm. I think you are right -- and return NO if not enabled.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
```diff
@@ -554,15 +554,34 @@ - (_WKWebsiteModalContainerObservationPolicy)_modalContainerObservationPolicy
+
+- (void)_setLockdownModeEnabled:(BOOL)lockdownModeEnabled
+{
+#if PLATFORM(IOS_FAMILY)
```
Shouldn't this be protected with `#if ENABLE(LOCKDOWN_MODE_API)`?

brentfulgham replied:
Will fix.

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
```diff
@@ -554,15 +554,34 @@ - (_WKWebsiteModalContainerObservationPolicy)_modalContainerObservationPolicy
-#if USE(APPLE_INTERNAL_SDK)
-#import <WebKitAdditions/WKWebpagePreferencesAdditions.mm>
-#else
+- (BOOL)lockdownModeEnabled
```
This should be `- (BOOL)isLockdownModeEnabled`, too.

brentfulgham replied:
Will fix.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
```diff
@@ -554,15 +554,34 @@ - (_WKWebsiteModalContainerObservationPolicy)_modalContainerObservationPolicy
-#if USE(APPLE_INTERNAL_SDK)
-#import <WebKitAdditions/WKWebpagePreferencesAdditions.mm>
-#else
+- (BOOL)lockdownModeEnabled
```
I discussed this with Tim and Wenson, it seems both our getter and setter are wrong here.

This one should be `isLockdownModeEnabled` and the setter should be `setLockdownModeEnabled`.

Could you please update all our API tests to use the API instead of the `_setCaptivePortalModeEnabled / _captivePortalModeEnabled` SPI, this would have caught the fact that our API doesn't work...

brentfulgham replied:
Yeah, let me give that a try.

brentfulgham replied:
@cdumez Should we remove (or deprecate) the SPI?

cdumez replied:
It is SPI so I think we can remove it in a follow-up, once we've ported clients to the new API.

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
```diff
@@ -8276,7 +8276,7 @@ void configureCaptivePortalWKWebViewConfiguration(WKWebViewConfiguration *config
     delegate.get().decidePolicyForNavigationActionWithPreferences = ^(WKNavigationAction *action, WKWebpagePreferences *preferences, void (^completionHandler)(WKNavigationActionPolicy, WKWebpagePreferences *)) {
         EXPECT_TRUE(preferences._captivePortalModeEnabled);
-        [preferences _setCaptivePortalModeEnabled:NO]; // Explicitly Opt out of captive portal mode for this load.
+        preferences.lockdownModeEnabled = YES; // Explicitly Opt out of captive portal mode for this load.
```
Looks like you reversed thing here.

brentfulgham replied:
Oh goodness. :-(

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
```diff
@@ -7906,7 +7906,7 @@ void configureCaptivePortalWKWebViewConfiguration(WKWebViewConfiguration *config
 
     delegate.get().decidePolicyForNavigationActionWithPreferences = ^(WKNavigationAction *action, WKWebpagePreferences *preferences, void (^completionHandler)(WKNavigationActionPolicy, WKWebpagePreferences *)) {
         EXPECT_FALSE(preferences._captivePortalModeEnabled);
```
There are a bunch of references to _captivePortalModeEnabled that remain in this file and that should be updated to use the new API too.

brentfulgham replied:
I'll do a more thorough search for instances of it.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
```diff
@@ -554,15 +554,40 @@ - (_WKWebsiteModalContainerObservationPolicy)_modalContainerObservationPolicy
+}
+
 - (void)_willChangeCaptivePortalMode
 {
```
This is supposed to go away

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
```diff
@@ -554,15 +554,40 @@ - (_WKWebsiteModalContainerObservationPolicy)_modalContainerObservationPolicy
+#endif
 }
+
 - (void)_didChangeCaptivePortalMode
```
ditto.

cdumez reviewed (approved):
LGTM with change.

brentfulgham committed:
Land Lockdown Mode features in Open Source
https://bugs.webkit.org/show_bug.cgi?id=242767

Reviewed by Chris Dumez.

* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
(-[WKWebpagePreferences lockdownModeEnabled]):
(-[WKWebpagePreferences _setLockdownModeEnabled:]):
(-[WKWebpagePreferences _willChangeCaptivePortalMode]):
(-[WKWebpagePreferences _didChangeCaptivePortalMode]):
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _presentCaptivePortalModeAlertOnce]):
(-[WKWebView _presentCaptivePortalModeAlertIfNeeded]):

Canonical link: https://commits.webkit.org/252540@main
